### PR TITLE
Provide the relation itself to relatedModel function handlers.

### PR DESCRIPTION
**Pull request by monokrome**
For one example of where this is useful, please see:

https://gist.github.com/monokrome/5954751

This happens in any case where a function is abstracted out and the relation data can not be assumed, so I would think that it would be a fairly common issue.

### CODE REVIEW

File: test/associated-model.js [hunk 1]
```diff
@@ -903,6 +907,8 @@ $(document).ready(function () {
         equal(cjob.get('organizable') instanceof Models.Company, true);
         equal(djob.get('organizable') instanceof Models.Department, true);
 
+        equal(relation, Models.Job.prototype.relations[0]);
```
Comment (sgrif) [inline test/associated-model.js]: This test doesn't provide any context on how we're expecting to use the relation, or what the immediate externally visible results would be.

Comment (monokrome) [inline test/associated-model.js]: The visible results would be that you have a reference to the expected relation, which is the goal of the feature. I'm not sure what more would need to be demonstrated in this case, but am happy to implement if this is not satisfactory.

Comment (sgrif) [inline test/associated-model.js]: Tests tend to act as a form of documentation for the code base, and I don't think the usage of this feature is clear from this test. How would you feel about an example in the tests to demonstrate what the actual intended use of passing this to the function would be, with a pseudo realistic use case (e.g. the assertions two lines above)

Comment (sgrif) [inline test/associated-model.js]: Or to be more specific, what do you think about a separate test that demonstrates what the model using this would look like, the data set that it would receive, and the state the relation would be in after making use of this?

Comment (monokrome) [inline test/associated-model.js]: I have made amendments to the code in order to make the use case of this feature more apparent. Does this represent more of what you expected from the test?

File: test/associated-model.js [hunk 2]
```diff
@@ -877,15 +877,17 @@ $(document).ready(function () {
...
+
         Models.Job = Backbone.AssociatedModel.extend({
             relations:[
                 {
                     type:Backbone.One,
                     key:'organizable',
-                    relatedModel:function (attributes) {
-                        return Models[attributes['organizable_type'] || this.get('organizable_type')];
-                    }
+                    relatedModel: findRelatedType
```
Comment (sgrif) [inline test/associated-model.js:891]: Maybe I'm missing something, but this doesn't seems to add additional complexity to handle the same case. What do you think about just having the consumer code use a simple macro if needed to reduce duplication?

``` javascript
function simplePolymorphicRelation(relationKey) {
  return function(attributes) {
     return Models[attributes[relationKey + '_type'] || this.get(relationKey + '_type')];
  };
};

Models.Job = Backbone.AssociatedModel.extend({
  relations: [
    {
      type: Backbone.One,
      key: 'organizable',
      relatedModel: simplePolymorphicRelation('organizable')
    }
  ]
});
```

Comment (monokrome) [inline test/associated-model.js:891]: @sgrif See here: https://gist.github.com/monokrome/5954751#comment-859142

Making the test reflect that would be more confusing than useful.

Comment (sgrif) [inline test/associated-model.js:891]: Maybe I'm still missing something, but that code doesn't appear to make any use of the relation at all. All it's doing is finding a relation by `relation.key == key` and using `relation.key` and nothing else. `key` could have been used by itself, as in the example in my previous comment.

Comment (monokrome) [inline test/associated-model.js:891]: Also, having to specify the same key to a function after having already specified it as "key" to the relation is represents a case where the library is causing the developer using the library to deal with a non-ideal interface due to a lack of support in the library.

The function that you provided might solve the problem, but that doesn't mean that it solves it well. Why duplicate the key when it is already in that configuration? Considering the fact that we know that this function was designed to work with some kind of relation, it would seem intuitive that a reference to the relation would be passed to the function.

Comment (jdkanani) [inline test/associated-model.js:891]: or even smaller and simple:

``` javascript
function simplePolymorphicRelation(attributes, relation) {
     var key = relation.key; // shows use of relation
     return Models[attributes[key+ '_type'] || this.get(key+ '_type')];
};

Models.Job = Backbone.AssociatedModel.extend({
  relations: [
    {
      type: Backbone.One,
      key: 'organizable',
      relatedModel: simplePolymorphicRelation
    }
  ]
});
```

Comment (monokrome) [inline test/associated-model.js:891]: @sgrif I linked you to the comment specifically mentioning that `key` would not be necessary, but I had no access to the relation as provided from the library. The fact that I don't have the value of the relation to get the `key` from it meant that I had to make the function twice as complicated as well as having to make it require the user provide extra information to the function.

@jdkanani Agreed. I'll update the code.

Comment (sgrif) [inline test/associated-model.js:891]: Fair enough. This just seems like an edge case to me. The example of `Array.map` when called with `parseInt` leading to unexpected results due to extra arguments comes to mind.

Comment (monokrome) [inline test/associated-model.js:891]: The `parseInt` issue is just that sometimes developers use `parseInt` with only one argument instead of two, and `parseInt` requires two. Developers don't define `parseInt`, it's an interface given to them. It's a much different issue for two reasons:
- The user has control of what the function is doing in this case, because they are writing it. In JavaScript this isn't as much of a point, but even in the parseInt example - you can wrap parseInt in a closure and pass the correct secondary argument to it and the problem is fixed. It's another case of user error - and a case that wouldn't happen with this example.
- JavaScript provides that all arguments are optional unless expected by the function (where an error is thrown), so not expecting `relation` to exist function signature does not hurt anyone.

There aren't really any downsides to this patch as far as I can tell in terms of it breaking any previously existing code (that's specifically why I provide it as the second argument), and it proves to be extremely useful in just my first 5 minutes of using these polymorphic relations.

Thus, I don't see any potential issues which could arrise from this code even in more obscure scenarios.

Comment (sgrif) [inline test/associated-model.js:891]: Again, my argument is simply the fact that we're adding complexity to the API for very little gain. The same can be accomplished with one line of code in the consumer. This feels like it's losing sight of the fact that this is supposed to be a declarative format for OOP. We've gone from an API taking 0 arguments, and only relying on the state of the object we're modifying, to something that takes all possible outside data, and completely ignores everything about the actual object we're operating on.

Comment (dhruvaray) [inline test/associated-model.js:891]: @sgrif 

The nice thing about @jdkanani 's code is that we can reuse that function across multiple relations in a `AssociatedModel`. More importantly, even across different `AssociatedModel` types. Passing in `relation` helps in this scenario. See the snippet below

``` javascript

function simplePolymorphicRelation(attributes, relation) {
     var key = relation.key; // shows use of relation
     return Models[attributes[key+ '_type'] || this.get(key+ '_type')];
};

Models.Job = Backbone.AssociatedModel.extend({
  relations: [
    {
      type: Backbone.One,
      key: 'organizable',
      relatedModel: simplePolymorphicRelation
    }
  ]
});

Models.Article = Backbone.AssociatedModel.extend({
  relations: [
    {
      type: Backbone.One,
      key: 'organizable',
      relatedModel: simplePolymorphicRelation
    }
  ]
});

```

Comment (monokrome) [inline test/associated-model.js:891]: I changed the code to reflect @jdkanani's change in the most recent update, by the way. Although, I did add the `_type` part to the variable to simplify.

Comment (sgrif) [inline test/associated-model.js:891]: @dhruvaray That can still be easily accomplished with the same simple macro in the consumer, without needing to add to the API. The difference is literally 1 line of code:

``` javascript
function simplePolymorphicRelation(relationKey) {
  return function(attributes) {
     return Models[attributes[relationKey + '_type'] || this.get(relationKey + '_type')];
  };
};

Models.Job = Backbone.AssociatedModel.extend({
  relations: [
    {
      type: Backbone.One,
      key: 'organizable',
      relatedModel: simplePolymorphicRelation('organizable')
    }
  ]
});

Models.Article = Backbone.AssociatedModel.extend({
  relations: [
    {
      type: Backbone.One,
      key: 'commentable',
      relatedModel: simplePolymorphicRelation('commentable')
    }
  ]
});
```

Comment (sgrif) [inline test/associated-model.js:891]: This really shouldn't be a huge deal, either way. I'll see if it's possible to remove the need for the attributes to be passed in, at the very least.

Comment (dhruvaray) [inline test/associated-model.js:891]: Yes... The real nasty bit about this API is passing `attributes`from outside.

Comment (monokrome) [inline test/associated-model.js:891]: As @sgrif mentioned, passing in `_.extend(_.clone(this.attributes), attributes)` should solve that problem. I'd say that's the most elegant solution for the current issues in #23. I don't think that it's safe to update the model at this point with the data yet (or Backbone would have, I'd think), so it probably makes sense to pass attributes into the model.

You could also attach attributes to the relation before passing it in, but I don't think that'd make much sense. Passing attributes is probably the most realistic way to solve the problem.

Comment (dhruvaray) [inline test/associated-model.js:891]: On the code snippet, there's probably no reason to clone. This should also do the trick...

``` javascript
_.extend({},this.attributes,attributes)
```

Comment (sgrif) [inline test/associated-model.js:891]: Heh. That's literally the exact implementation of `_.clone`

On Tue, Jul 9, 2013 at 12:31 AM, Dhruva Ray notifications@github.comwrote:

> In test/associated-model.js:
> 
> > ```
> >      Models.Job = Backbone.AssociatedModel.extend({
> >          relations:[
> >              {
> >                  type:Backbone.One,
> >                  key:'organizable',
> > ```
> > -                    relatedModel:function (attributes) {
> > -                        return Models[attributes['organizable_type'] || this.get('organizable_type')];
> > -                    }
> > -                    relatedModel: findRelatedType
> 
> On the code snippet, there's probably no reason to clone.
> 
> _.extend({},this.attributes,attributes)
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/dhruvaray/backbone-associations/pull/50/files#r5083032
> .

## 

Thanks,
Sean Griffin

Comment (monokrome) [inline test/associated-model.js:891]: The `_.clone` function is pretty useless overall. In large data sets, those extra function calls can add up and it isn't really doing anything complicated.

On a side note, what's the process for getting this pull request accepted or rejected? Any other changes needed, or are there just other eyes that need to take a look at it before a decision can be made?

**Comment by monokrome** [comment]
Continuation of [this comment](https://github.com/dhruvaray/backbone-associations/pull/50#discussion_r5083069) should probably occur over here.

File: test/associated-model.js (hunk 2 above)
Comment (dhruvaray) [inline test/associated-model.js:891]: @sgrif : You are right... _.clone is a shallow one and is the same.

Comment (dhruvaray) [inline test/associated-model.js:891]: @monokrome : We will take in the pull request. But if @sgrif comes up with a way to avoid the attributes or a different approach, this may change.

**Comment by dhruvaray** [comment]
Would be good to switch the order of the parameters around - relation, attributes. (Rather than attributes, relation). Then we should be good for a merge.

**Comment by dhruvaray** [comment]
I have updated the test case and changed order of parameters -  ad43c85
